File: tests/wasm/test_sqwma.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('SQWMA partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.sqwma_js(close, 14);
    assert.strictEqual(result.length, close.length);
});

test('SQWMA accuracy', async () => {

    const close = new Float64Array(testData.close);


    const expectedLastFive = [
        59229.72287968442,
        59211.30867850099,
        59172.516765286,
        59167.73471400394,
        59067.97928994083,
    ];

    const result = wasm.sqwma_js(close, 14);

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expectedLastFive,
        1e-8,
        "SQWMA last 5 values mismatch"
    );


    await compareWithRust('sqwma', result, 'close', { period: 14 });
});

test('SQWMA zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.sqwma_js(inputData, 0);
    }, /Invalid period/);
});

test('SQWMA period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.sqwma_js(dataSmall, 10);
    }, /Invalid period/);
});

test('SQWMA very small dataset', () => {

    const singlePoint = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.sqwma_js(singlePoint, 9);
    }, /Invalid period/);
});

test('SQWMA empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.sqwma_js(empty, 14);
    }, /Input data slice is empty/);
});

test('SQWMA NaN handling', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.sqwma_js(close, 14);
    assert.strictEqual(result.length, close.length);


    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }



    for (let i = 0; i < 15; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} in warmup period`);
    }
});

test('SQWMA batch single period', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.sqwma_batch_js(close, 14, 14, 0);


    assert.strictEqual(result.length, close.length);


    const expectedLastFive = [
        59229.72287968442,
        59211.30867850099,
        59172.516765286,
        59167.73471400394,
        59067.97928994083,
    ];


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expectedLastFive,
        1e-8,
        "SQWMA batch default row mismatch"
    );
});

test('SQWMA batch metadata', () => {

    const periods = wasm.sqwma_batch_metadata_js(10, 20, 2);


    const expectedPeriods = [10, 12, 14, 16, 18, 20];
    assert.deepStrictEqual(Array.from(periods), expectedPeriods);
});

test('SQWMA all NaN input', () => {

    const allNan = new Float64Array(100).fill(NaN);

    assert.throws(() => {
        wasm.sqwma_js(allNan, 14);
    }, /All values are NaN/);
});

test('SQWMA period less than 2', () => {

    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);

    assert.throws(() => {
        wasm.sqwma_js(data, 1);
    }, /Invalid period/);
});

test('SQWMA batch multiple periods', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.sqwma_batch_js(close, 10, 20, 2);


    assert.strictEqual(result.length, 6 * close.length);


    const periods = wasm.sqwma_batch_metadata_js(10, 20, 2);
    assert.strictEqual(periods.length, 6);


    for (let i = 0; i < 6; i++) {
        const period = periods[i];
        const startIdx = i * close.length;
        const rowData = result.slice(startIdx, startIdx + close.length);


        const warmupEnd = period + 1;
        for (let j = 0; j < warmupEnd; j++) {
            assert(isNaN(rowData[j]), `Expected NaN in warmup for period ${period} at index ${j}`);
        }


        if (rowData.length > warmupEnd + 10) {
            let hasValidValues = false;
            for (let j = warmupEnd; j < warmupEnd + 10; j++) {
                if (!isNaN(rowData[j])) {
                    hasValidValues = true;
                    break;
                }
            }
            assert(hasValidValues, `Expected valid values after warmup for period ${period}`);
        }
    }
});

test('SQWMA edge case period 2', () => {

    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);

    const result = wasm.sqwma_js(data, 2);
    assert.strictEqual(result.length, data.length);


    assertAllNaN(result.slice(0, 3));


    assertNoNaN(result.slice(3));
});

test('SQWMA batch with step 0', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.sqwma_batch_js(close, 15, 15, 0);
    assert.strictEqual(result.length, close.length);


    const periods = wasm.sqwma_batch_metadata_js(15, 15, 0);
    assert.deepStrictEqual(Array.from(periods), [15]);
});

test('SQWMA consistency check', () => {

    const close = new Float64Array(testData.close);
    const period = 14;

    const single = wasm.sqwma_js(close, period);
    const batch = wasm.sqwma_batch_js(close, period, period, 0);

    assert.strictEqual(single.length, batch.length);



    for (let i = 0; i < single.length; i++) {
        if (isNaN(single[i]) && isNaN(batch[i])) {
            continue;
        }




        assertClose(single[i], batch[i], 2e-9, 2e-9,
            `SQWMA single vs batch mismatch at index ${i}`);
    }
});

test('SQWMA with leading NaNs', () => {


    const data = new Float64Array(15);
    for (let i = 0; i < 5; i++) {
        data[i] = NaN;
    }
    for (let i = 5; i < 15; i++) {
        data[i] = i - 4;
    }

    const result = wasm.sqwma_js(data, 3);
    assert.strictEqual(result.length, data.length);


    for (let i = 0; i < 9; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} in warmup period including leading NaNs`);
    }

    for (let i = 9; i < result.length; i++) {
        assert(!isNaN(result[i]), `Expected valid value at index ${i} after warmup`);
    }
});

test('SQWMA unified batch API', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const batchResult = wasm.sqwma_batch(close, {
        period_range: [10, 20, 5]
    });


    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);


    const periods = [10, 15, 20];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.values.slice(rowStart, rowEnd);

        const singleResult = wasm.sqwma_js(close, periods[i]);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} mismatch in unified batch API`
        );
    }
});

test('SQWMA batch metadata from result', () => {

    const close = new Float64Array(30);
    close.fill(100);

    const result = wasm.sqwma_batch(close, {
        period_range: [10, 20, 10]
    });


    assert.strictEqual(result.combos.length, 2);


    assert.strictEqual(result.combos[0].period, 10);


    assert.strictEqual(result.combos[1].period, 20);
});

test('SQWMA improved warmup assertions', () => {

    const data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        data[i] = i + 1;
    }


    const testCases = [
        { period: 5, warmupEnd: 6 },
        { period: 10, warmupEnd: 11 },
        { period: 15, warmupEnd: 16 },
    ];

    for (const { period, warmupEnd } of testCases) {
        const result = wasm.sqwma_js(data, period);


        for (let i = 0; i < warmupEnd; i++) {
            assert(isNaN(result[i]),
                `Period ${period}: Expected NaN at index ${i} (warmup ends at ${warmupEnd})`);
        }


        for (let i = warmupEnd; i < Math.min(warmupEnd + 5, result.length); i++) {
            assert(!isNaN(result[i]),
                `Period ${period}: Expected valid value at index ${i} (warmup ended at ${warmupEnd})`);
        }
    }
});

test('SQWMA batch with NaN injection', () => {

    const data = new Float64Array(30);
    for (let i = 0; i < 30; i++) {
        data[i] = i + 1;
    }

    data[5] = NaN;
    data[6] = NaN;

    const result = wasm.sqwma_batch_js(data, 5, 10, 5);


    const periods = wasm.sqwma_batch_metadata_js(5, 10, 5);
    assert.strictEqual(periods.length, 2);




    for (let p = 0; p < 2; p++) {
        const period = periods[p];
        const rowStart = p * data.length;
        const rowData = result.slice(rowStart, rowStart + data.length);




        const theoreticalEarliest = 7 + (period - 2);


        for (let i = 0; i < theoreticalEarliest; i++) {
            assert(
                isNaN(rowData[i]),
                `Period ${period}: Expected NaN at index ${i} (pre-window-clean)`
            );
        }


        let firstFinite = -1;
        for (let i = theoreticalEarliest; i < rowData.length; i++) {
            if (!isNaN(rowData[i])) { firstFinite = i; break; }
        }
        if (firstFinite !== -1) {

            for (let i = firstFinite; i < Math.min(firstFinite + 3, rowData.length); i++) {
                assert(!isNaN(rowData[i]),
                    `Period ${period}: Expected valid value at index ${i} (first finite at ${firstFinite})`);
            }
        }
    }
});

File: tests/python/test_sqwma.py
"""
Python binding tests for SQWMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    # If not in virtual environment, try to import from installed location
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestSqwma:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_sqwma_partial_params(self, test_data):
        """Test SQWMA with partial parameters (None values) - mirrors check_sqwma_partial_params"""
        close = test_data['close']
        
        # Test with default period (14)
        result = ta_indicators.sqwma(close, 14)
        assert len(result) == len(close)
    
    def test_sqwma_accuracy(self, test_data):
        """Test SQWMA matches expected values from Rust tests - mirrors check_sqwma_accuracy"""
        close = test_data['close']
        
        # Expected values from Rust test
        expected_last_five = [
            59229.72287968442,
            59211.30867850099,
            59172.516765286,
            59167.73471400394,
            59067.97928994083,
        ]
        
        result = ta_indicators.sqwma(close, period=14)
        
        assert len(result) == len(close)
        
        # Add debug output
        print(f"\nSQWMA test_accuracy debug:")
        print(f"Input length: {len(close)}")
        print(f"Output length: {len(result)}")
        print(f"Actual last 5: {result[-5:]}")
        print(f"Expected last 5: {expected_last_five}")
        
        # Check last 5 values match expected
        assert_close(
            result[-5:], 
            expected_last_five,
            rtol=1e-8,
            msg="SQWMA last 5 values mismatch"
        )
        
        # Compare full output with Rust
        compare_with_rust('sqwma', result, 'close', {'period': 14})
    
    def test_sqwma_zero_period(self):
        """Test SQWMA fails with zero period - mirrors check_sqwma_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.sqwma(input_data, period=0)
    
    def test_sqwma_period_exceeds_length(self):
        """Test SQWMA fails when period exceeds data length - mirrors check_sqwma_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.sqwma(data_small, period=10)
    
    def test_sqwma_very_small_dataset(self):
        """Test SQWMA fails with insufficient data - mirrors check_sqwma_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.sqwma(single_point, period=9)
    
    def test_sqwma_nan_handling(self, test_data):
        """Test SQWMA handles NaN values correctly - mirrors check_sqwma_nan_handling"""
        close = test_data['close']
        period = 14
        
        result = ta_indicators.sqwma(close, period=period)
        assert len(result) == len(close)
        
        # Find first non-NaN in input data
        first_valid = next((i for i, v in enumerate(close) if not np.isnan(v)), 0)
        
        # SQWMA warmup period: first_non_nan + period + 1
        warmup_end = first_valid + period + 1
        
        # Check warmup period has NaN values
        assert np.all(np.isnan(result[:warmup_end])), f"Expected NaN in warmup period (indices 0-{warmup_end-1})"
        
        # After warmup period, no NaN values should exist (if we have enough data)
        if len(result) > warmup_end:
            # Check a reasonable range after warmup
            check_start = min(warmup_end, 240)
            if len(result) > check_start:
                assert not np.any(np.isnan(result[check_start:])), f"Found unexpected NaN after index {check_start}"
    
    def test_sqwma_streaming(self, test_data):
        """Test SQWMA streaming matches batch calculation - mirrors check_sqwma_streaming"""
        close = test_data['close']
        period = 14
        
        # Batch calculation
        batch_result = ta_indicators.sqwma(close, period=period)
        
        # Streaming calculation
        stream = ta_indicators.SqwmaStream(period=period)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        # Compare batch vs streaming
        assert len(batch_result) == len(stream_values)
        
        # Compare values where both are not NaN
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9, 
                        msg=f"SQWMA streaming mismatch at index {i}")
    
    def test_sqwma_batch(self, test_data):
        """Test SQWMA batch processing - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.sqwma_batch(
            close,
            period_range=(14, 14, 0),  # Default period only
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        # Should have 1 combination (default params)
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        
        # Extract the single row
        default_row = result['values'][0]
        expected_last_five = [
            59229.72287968442,
            59211.30867850099,
            59172.516765286,
            59167.73471400394,
            59067.97928994083,
        ]
        
        # Check last 5 values match
        assert_close(
            default_row[-5:],
            expected_last_five,
            rtol=1e-8,
            msg="SQWMA batch default row mismatch"
        )
    
    def test_sqwma_empty_input(self):
        """Test SQWMA fails with empty input - mirrors check_sqwma_empty_input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.sqwma(empty, period=14)
    
    def test_sqwma_all_nan_input(self):
        """Test SQWMA with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.sqwma(all_nan, period=14)
    
    def test_sqwma_period_less_than_2(self):
        """Test SQWMA fails with period < 2"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.sqwma(data, period=1)
    
    def test_sqwma_kernel_selection(self, test_data):
        """Test SQWMA with different kernel selections"""
        close = test_data['close']
        
        # Test scalar kernel
        result_scalar = ta_indicators.sqwma(close, period=14, kernel='scalar')
        assert len(result_scalar) == len(close)
        
        # Test auto kernel
        result_auto = ta_indicators.sqwma(close, period=14, kernel='auto')
        assert len(result_auto) == len(close)
        
        # Test invalid kernel
        with pytest.raises(ValueError, match="Unknown kernel"):
            ta_indicators.sqwma(close, period=14, kernel='invalid_kernel')
    
    def test_sqwma_batch_multiple_periods(self, test_data):
        """Test SQWMA batch with multiple periods"""
        close = test_data['close']
        
        result = ta_indicators.sqwma_batch(
            close,
            period_range=(10, 20, 2),  # periods: 10, 12, 14, 16, 18, 20
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        # Should have 6 combinations
        assert result['values'].shape[0] == 6
        assert result['values'].shape[1] == len(close)
        
        # Check periods array
        expected_periods = [10, 12, 14, 16, 18, 20]
        assert np.array_equal(result['periods'], expected_periods)
        
        # Verify each row has appropriate warmup period
        for i, period in enumerate(expected_periods):
            row = result['values'][i]
            # Check that we have NaN values in the warmup period (period + 1)
            warmup_end = period + 1
            assert np.all(np.isnan(row[:warmup_end])), f"Expected NaN in warmup for period {period}"
            # Check that we have valid values after warmup (if data is long enough)
            if len(row) > warmup_end + 10:
                assert not np.all(np.isnan(row[warmup_end:warmup_end+10])), f"Expected valid values after warmup for period {period}"
    
    def test_sqwma_edge_case_period_2(self):
        """Test SQWMA with minimum valid period (2)"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        
        result = ta_indicators.sqwma(data, period=2)
        assert len(result) == len(data)
        
        # First 3 values (period + 1) should be NaN
        assert np.all(np.isnan(result[:3]))
        # Remaining values should be valid
        assert not np.any(np.isnan(result[3:]))
    
    def test_sqwma_with_leading_nans(self):
        """Test SQWMA correctly handles data that starts with NaN values"""
        # Create data with 5 leading NaNs
        data = np.array([np.nan] * 5 + [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0])
        
        result = ta_indicators.sqwma(data, period=3)
        assert len(result) == len(data)
        
        # First non-NaN is at index 5, so warmup ends at 5 + 3 + 1 = 9
        assert np.all(np.isnan(result[:9])), "Expected NaN in warmup period including leading NaNs"
        # Should have valid values starting from index 9
        assert not np.any(np.isnan(result[9:])), "Expected valid values after warmup"
        
        # Test with different kernels to ensure they all handle it correctly
        for kernel in ['scalar', 'auto']:
            result_k = ta_indicators.sqwma(data, period=3, kernel=kernel)
            assert np.all(np.isnan(result_k[:9])), f"Kernel {kernel}: Expected NaN in warmup"
            assert not np.any(np.isnan(result_k[9:])), f"Kernel {kernel}: Expected valid values after warmup"
    
    def test_sqwma_boundary_conditions(self):
        """Test SQWMA with boundary conditions"""
        # Test with exact minimum data (period = data length)
        period = 5
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        
        result = ta_indicators.sqwma(data, period=period)
        assert len(result) == len(data)
        # All should be NaN since warmup = 0 + period + 1 = 6, which exceeds data length
        assert np.all(np.isnan(result)), "Expected all NaN when data length equals period"
        
        # Test with just enough data for one output
        data_extended = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0])  # length = 7
        period = 3  # warmup = 0 + 3 + 1 = 4, so we get output at indices 4,5,6
        
        result = ta_indicators.sqwma(data_extended, period=period)
        assert len(result) == len(data_extended)
        assert np.all(np.isnan(result[:4])), "Expected NaN in warmup period"
        assert not np.any(np.isnan(result[4:])), "Expected valid values after warmup"
        
        # Test with period = 2 (minimum valid period)
        data_min = np.array([1.0, 2.0, 3.0, 4.0])
        result = ta_indicators.sqwma(data_min, period=2)
        assert len(result) == len(data_min)
        # Warmup = 0 + 2 + 1 = 3
        assert np.all(np.isnan(result[:3])), "Expected NaN in first 3 values for period=2"
        assert not np.isnan(result[3]), "Expected valid value at index 3 for period=2"
    
    def test_sqwma_batch_parameter_sweep(self, test_data):
        """Test SQWMA batch with comprehensive parameter sweep"""
        close = test_data['close'][:100]  # Use smaller dataset for speed
        
        # Test with multiple period combinations
        result = ta_indicators.sqwma_batch(
            close,
            period_range=(5, 25, 5),  # periods: 5, 10, 15, 20, 25
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        # Should have 5 combinations
        assert result['values'].shape[0] == 5
        assert result['values'].shape[1] == len(close)
        
        # Check periods array
        expected_periods = [5, 10, 15, 20, 25]
        assert np.array_equal(result['periods'], expected_periods)
        
        # Verify each row against single calculation
        for i, period in enumerate(expected_periods):
            row = result['values'][i]
            single_result = ta_indicators.sqwma(close, period=period)
            
            # Compare where both are valid
            for j in range(len(row)):
                if np.isnan(row[j]) and np.isnan(single_result[j]):
                    continue
                assert_close(row[j], single_result[j], rtol=1e-10, atol=1e-10,
                           msg=f"Batch vs single mismatch for period {period} at index {j}")


if __name__ == '__main__':
    pytest.main([__file__, '-v'])